DataFormats/Common/interface/DetSetVectorNew.h

```diff
@@ -212,13 +275,106 @@ namespace edmNew {
       typename DetSetVector<T>::Item & item;
       bool saveEmpty;
     };
+
+    /* fill on demand a given  DetSet
+     */
+    class TSFastFiller {
+    public:
+      typedef typename DetSetVector<T>::data_type value_type;
+      typedef typename DetSetVector<T>::id_type key_type;
+      typedef typename DetSetVector<T>::id_type id_type;
+      typedef typename DetSetVector<T>::size_type size_type;
+
+#ifdef USE_ATOMIC
+      static DetSetVector<T>::Item & dummy() {
+        static  DetSetVector<T>::Item d; return d;
+      }
+      TSFastFiller(DetSetVector<T> & iv, id_type id) :
+        v(iv), item(v.ready()? v.push_back(id): dummy()) { assert(v.filling==true); v.filling = false;}
+
+      TSFastFiller(DetSetVector<T> & iv, typename DetSetVector<T>::Item & it) : 
+	v(iv), item(it) {
+
+      }
+      ~TSFastFiller() {
+	bool expected=false;
+	while (!v.filling.compare_exchange_weak(expected,true,std::memory_order_acq_rel))  { expected=false; nanosleep(0,0);}
+	int offset = v.m_data.size();
+	if (v.m_data.capacity()<offset+lv.size()) {
+          /// throw?? will throw the one below but maybe better to throw here...
+        }
+	std::move(lv.begin(), lv.end(), std::back_inserter(v.m_data));
+	item.size=lv.size();
+	item.offset = offset; 
+
+	assert(v.filling==true);
+	v.filling.store(false,std::memory_order_release);
+      }
+      
+#endif
+      
+      void abort() {
+	lv.clear();
+      }
+
+      void reserve(size_type s) {
+	lv.reserve(s);
+      }
+
+      void resize(size_type s) {
+	lv.resize(s);
+      }
+
+      id_type id() const { return item.id;}
+      size_type size() const { return lv.size();}
+      bool empty() const { return lv.empty();}
+
+      data_type & operator[](size_type i) {
+	return 	lv[i];
+      }
+      DataIter begin() { return lv.begin();}
+      DataIter end() { return lv.end();}
+
+      void push_back(data_type const & d) {
+	lv.push_back(d);
+      }
+#ifndef CMS_NOCXX11
+      void push_back(data_type && d) {
+        lv.push_back(std::move(d));
+      }
+#endif
+
+      data_type & back() { return v.m_data.back();}
```

Comment:
Shouldn't this be ```lv.back()```? If not then this isn't thread safe since ```v.m_data``` is not being protected by ```v.filler``` during this call nor is the return value guaranteed to be properly initialized without that synchronization.